ra-AlexWolfMSFT-pr Scalable web apps

alexwolfmsft opened the pull request:
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps?view=aspnetcore-7.0&branch=pr-en-us-26956&tabs=login-azure-cli)

Rick-Anderson commented:
@alexwolfmsft  way too many screen shots. The Azure UI changes hourly and we don't have the resources to keep up. Better to just have instructions and a few screen shots. It's much easier to do a quick update on markup instructions than reshooting a bunch of images.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,352 @@
+1. [Connect the Azure services](#5-connect-the-azure-services)
+1. [Configure and redeploy the app](#6-configure-and-redeploy-the-app)
+
+In some cases, simple ASP.NET Core apps are able to scale without special considerations. However, some apps require additional configurations to utilize certain framework features or architectural patterns, including the following:
```
simple is banned from this repo. Please look at the changes I made rather than tossing all my work.

alexwolfmsft replied:
Thanks for the comments, this is my first time working in this repo. I moved the updated version of the article into this branch and am still updating it as a draft PR - I started integrating your recommendations and will go through the rest as well. I opened a draft PR to generate a preview as I work on this - is this the desired workflow for this repo?

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
````diff
@@ -0,0 +1,352 @@
+git clone "https://github.com/MicrosoftDocs/mslearn-dotnet-debug-visual-studio-app-service.git"
+```
+
+The sample app uses a search form to browse GitHub repositories by name. The form relies on the built-in ASP.NET Core data protection services to handle anti-forgery concerns. By default, when the app scales horizontally on Container Apps, the data protection service will throw an exception. You'll explore and solve this issue in the steps ahead.
````
Check out the MS style guide, FT is not allowed without an exception.

alexwolfmsft replied:
I'm probably missing something obvious here, but I'm not following - what is FT?

Rick-Anderson replied:
Sorry, Future Tense, look for `will`

Rick-Anderson replied:
the data protection service ~will~ throws an exception

MT (machine translates) much better, is more clear, less words

alexwolfmsft replied:
Done

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,352 @@
+
+:::image type="content" source="./media/scaling-app-screenshot.png" alt-text="A screenshot showing the GitHub Explorer app.":::
+
+## 2) Deploy the app to Azure Container Apps
```
Don't put numbers in headings. If you're really concerned folks won't follow this in order, tell them to follow in order.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,352 @@
+
+## 2) Deploy the app to Azure Container Apps
+
+Next you'll use Visual Studio to deploy the app to Azure Container Apps. Container apps provide a managed service designed to simplify hosting containerized apps and microservices.
```
Next is banned.

alexwolfmsft replied:
Will fix/remember for the future.

alexwolfmsft commented:
Hey @Rick-Anderson - a summary of the changes:
- Integrated your markdown feedback into the new version of the doc. 
- Removed most of the images and kept/updated only the ones that feel most helpful to the user. 
- Moved the code project into the samples repo added code snippet samples where necessary 
- Fixed packages
- Other small verbiage/acronlinx updates as well. 

Thanks for the assistance on this.

alexwolfmsft commented:
@Rick-Anderson While I'm making these final tweaks - where would you like to see this on the TOC? Just realized that also needs to be updated. I was thinking it could go under the Host and Deploy section somewhere, with a title such as "Scaling apps on Azure" or "Scaling considerations" or something.

Rick-Anderson commented:
> where would you like to see this on the TOC?   was thinking it could go under the Host and Deploy section somewhere, with a title such as "Scaling apps on Azure" or "Scaling considerations" or something

I'll do that in another PR. In addition to Host and Deploy section, I can put it under Performance if you agree.

> Also, are we good to publish after the requested changes are set? I'm not sure if I'm supposed to be the one to merge or not for this repo.

Let me do a review, then once we agree on changes, I'll merge, add it to the TOC, then make it live.

alexwolfmsft commented:
> > where would you like to see this on the TOC?   was thinking it could go under the Host and Deploy section somewhere, with a title such as "Scaling apps on Azure" or "Scaling considerations" or something
> 
> I'll do that in another PR. In addition to Host and Deploy section, I can put it under Performance if you agree.
> 
> > Also, are we good to publish after the requested changes are set? I'm not sure if I'm supposed to be the one to merge or not for this repo.
> 
> Let me do a review, then once we agree on changes, I'll merge, add it to the TOC, then make it live.

That all sounds good thanks.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+        * Select **Ok** to close the container apps environment dialog.
+    * Select **Create** to close the original container apps dialog and Visual Studio creates the container app resource in Azure.
+1. Once the resource is created, make sure it's selected in the list of container apps, and then choose **Next**.
+1. You'll need to create an Azure Container Registry to store the published image artifact for your app. Select the green **+** icon on the container registry screen. Leave the default values, and then select **Create**.
```
```suggestion
1. Create an Azure Container Registry to store the published image artifact for the app. Select the green **+** icon on the container registry screen. Leave the default values, and then select **Create**.
```

Rick-Anderson replied:
Isn't it this?
<img width="67" alt="image" src="https://user-images.githubusercontent.com/3605364/190838194-3cb88aba-d9f6-4450-a2c7-dd4757070183.png">

I think we should write, Select **Create new**. You could even add an image for that

alexwolfmsft replied:
The registry is created in Visual Studio - this is the Azure Portal UI. Visual Studio has a green + icon and a step to create the registry after you've chosen container apps.

alexwolfmsft replied:
Woah nevermind, looks like the Visual Studio UI changed in the latest version, looks different for me now, I'll update.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    :::image type="content" source="./media/scaling-new-registry-small.png" lightbox="./media/scaling-new-registry.png" alt-text="A screenshot showing how to create a new container registry.":::
+
+1. After the container registry is created, make sure it's selected, and then choose finish to close the dialog workflow and display a summary of the publishing profile.
+1. Select **Publish** in the upper right of the publishing profile summary to deploy your app to Azure.
```
```suggestion
1. Select **Publish** in the upper right of the publishing profile summary to deploy the app to Azure.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. After the container registry is created, make sure it's selected, and then choose finish to close the dialog workflow and display a summary of the publishing profile.
+1. Select **Publish** in the upper right of the publishing profile summary to deploy your app to Azure.
+
+When the deployment finishes, Visual Studio launches the browser to display your hosted app. Search for `Microsoft` in the form field, and you should see a list of repositories displayed.
```
```suggestion
When the deployment finishes, Visual Studio launches the browser to display the hosted app. Search for `Microsoft` in the form field, and a list of repositories is displayed.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+## Scale and troubleshoot the app
+
+Your app is currently running without any issues, but you'd like to scale the app across more instances in anticipation of high traffic volumes.
```
```suggestion
The app is currently working without any issues, but we'd like to scale the app across more instances in anticipation of high traffic volumes.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. In the Azure portal, search for the `razorscaling-app-****` container app in the top level search bar and select it from the results.
+1. On the overview page, select **Scale** from the left navigation, and then select **+ Edit and deploy**.
+1. On the revisions page, switch to the **Scale** tab.
+1. Set both the min and max instances to **4** and then select **Create**. This configuration change guarantees your app is scaled horizontally across several instances.
```
several or 4?

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. On the revisions page, switch to the **Scale** tab.
+1. Set both the min and max instances to **4** and then select **Create**. This configuration change guarantees your app is scaled horizontally across several instances.
+
+Navigate back to your app in the browser. When the page loads, at first it appears everything is working correctly. However, when you enter in a search term again and hit submit, an error occurs. If you don't see the error at first, submit the form several more times.
```
```suggestion
Navigate back to the app. When the page loads, at first it appears everything is working correctly. However, when  a search term is entered and submitted, an error occurs. If an error is not displayed, submit the form several more times.
```
 Do we want to say. If the same instance responds to the form submission, no error occurs.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+It's not immediately apparent why the search requests are failing. The browser tools indicate a 400 Bad Request response was sent back. However, you can use the logging features of container apps to diagnose errors occurring in your environment.
+
+1. On the overview page of your container app, select **Logs** from the left navigation.
```
```suggestion
1. On the overview page of the container app, select **Logs** from the left navigation.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+1. On the overview page of your container app, select **Logs** from the left navigation.
+1. On the **Logs** page, close the pop-up that opens and navigate to the **Tables** tab.
+1. Expand the **Custom Logs** item to reveal the **ContainerAppConsoleLogs_CL** node. This table holds various logs for the container app that you can query to troubleshoot issues or questions.
```
```suggestion
1. Expand the **Custom Logs** item to reveal the **ContainerAppConsoleLogs_CL** node. This table holds various logs for the container app that can be queried to troubleshoot problems.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
````diff
@@ -0,0 +1,320 @@
+    | project ContainerAppName_s, Log_s
+    ```
+
+    This query searches the **ContainerAppConsoleLogs_CL** table for any rows that contain the word exception. The results are ordered by the time generated, limited to 500 results, and only include the **ContainerAppName_s** and **Log_s** columns to make the results easier to read.
````
```suggestion
    The preceding query searches the **ContainerAppConsoleLogs_CL** table for any rows that contain the word exception. The results are ordered by the time generated, limited to 500 results, and only include the **ContainerAppName_s** and **Log_s** columns to make the results easier to read.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+    This query searches the **ContainerAppConsoleLogs_CL** table for any rows that contain the word exception. The results are ordered by the time generated, limited to 500 results, and only include the **ContainerAppName_s** and **Log_s** columns to make the results easier to read.
+
+1. Select **Run**, and you should see a list of results printed. Read through the logs and note that most of them are related to antiforgery tokens and cryptography.
```
```suggestion
1. Select **Run**,  a list of results is displayed. Read through the logs and note that most of them are related to antiforgery tokens and cryptography.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    :::image type="content" source="./media/scaling-troubleshoot-small.png" lightbox="./media/scaling-troubleshoot.png" alt-text="A screenshot showing the logs query.":::
+
+    > [!IMPORTANT]
+    > The errors in the app are caused by the .NET data protection services. When multiple instances of the app are running, there is no guarantee that the HTTP POST request to submit the form is routed to the same container that initially loaded the page from the HTTP GET request. If the requests are handled by different instances, the antiforgery tokens cannot be handled correctly and an except ...
```
```suggestion
    > The errors in the app are caused by the .NET data protection services. When multiple instances of the app are running, there is no guarantee that the HTTP POST request to submit the form is routed to the same container that initially loaded the page from the HTTP GET request. If the requests are handled by different instances, the antiforgery tokens aren't handled correctly and an exception occurs.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    > [!IMPORTANT]
+    > The errors in the app are caused by the .NET data protection services. When multiple instances of the app are running, there is no guarantee that the HTTP POST request to submit the form is routed to the same container that initially loaded the page from the HTTP GET request. If the requests are handled by different instances, the antiforgery tokens cannot be handled correctly and an except ...
+
+    In the steps ahead, you'll resolve this issue by centralizing the data protection keys in an Azure storage service and protecting them with Key Vault.
```
```suggestion
    In the steps ahead, this issue is resolved by centralizing the data protection keys in an Azure storage service and protecting them with Key Vault.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+## Create the Azure Services
+
+To resolve the errors impacting the container app, you'll create the following services and connect them to your app:
```
```suggestion
To resolve the preceding errors,  the following services are created and connected to the app:
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+To resolve the errors impacting the container app, you'll create the following services and connect them to your app:
+
+* [Azure Storage Account](/azure/storage/common/storage-account-overview): Handles storing data for the Data Protection Services. Provides a centralized location to store key data as the app scales. Storage accounts can also be used to hold documents, queue data, file shares, and almost any type of blob data.
+* [Azure KeyVault](/azure/key-vault/general/basic-concepts): This service stores secrets for your app, and is used to help manage encryption concerns for the Data Protection Services.
```
```suggestion
* [Azure KeyVault](/azure/key-vault/general/basic-concepts): This service stores secrets for an app, and is used to help manage encryption concerns for the Data Protection Services.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Subscription**: Select the same subscription that you chose for the container app.
+    * **Resource Group**: Select the *msdocs-scalable-razor* resource group you created previously.
+    * **Storage account name**: Name the account *scalablerazorstorageXXXX* where the X's are random numbers of your choosing. This name must be unique across all of Azure.
+    * **Region**: Select a region that is close to you.
```
```suggestion
    * **Region**: Select the same region you previously selected.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Resource Group**: Select the *msdocs-scalable-razor* resource group you created previously.
+    * **Storage account name**: Name the account *scalablerazorstorageXXXX* where the X's are random numbers of your choosing. This name must be unique across all of Azure.
+    * **Region**: Select a region that is close to you.
+1. Leave the rest of the values at their default and select **Review**. After Azure validates your inputs, select **Create**.
```
```suggestion
1. Leave the rest of the values at their default and select **Review**. After Azure validates the inputs, select **Create**.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+#### Create the storage container
+
+You'll need to create a Container to store your app's data protection keys.
```
```suggestion
Create a Container to store the app's data protection keys.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Select **+ Add container** to open the **New container** flyout menu.
+1. Enter a name of *scalablerazorkeys*, leave the rest of the settings at their defaults, and then choose **Create**.
+
+You should see the new containers appear on the page list.
```
```suggestion
The new containers appear on the page list.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. In the Azure portal search bar, enter `Key Vault` and select the matching result.
+1. On the key vault listing page, select **+ Create**.
+1. On the **Basics** tab, enter the following values:
+    * **Subscription**: Select the same subscription that chose for your Container App.
```
```suggestion
    * **Subscription**: Select the same subscription that was previously selected.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. On the key vault listing page, select **+ Create**.
+1. On the **Basics** tab, enter the following values:
+    * **Subscription**: Select the same subscription that chose for your Container App.
+    * **Resource Group**: Select the *msdocs-scalable-razor* resource group you created previously.
```
```suggestion
    * **Resource Group**: Select the *msdocs-scalable-razor* resource group previously created .
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Subscription**: Select the same subscription that chose for your Container App.
+    * **Resource Group**: Select the *msdocs-scalable-razor* resource group you created previously.
+    * **Key Vault name**: Enter the name *scalablerazorvaultXXXX*.
+    * **Region**: Select a region near your location.
```
same as previous region?

alexwolfmsft replied:
I thought this might get confusing depending on which service they're creating and which are available in which regions at the time of reading, so I left this verbiage and added a note further up about choosing locations.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+1. On the main overview page of the key vault, select **Keys** from the left navigation.
+1. On the **Create a key** page, select **+ Generate/Import** to open the **Create a key** flyout menu. 
+1. Enter *razorkey* in the **Name** field. Leave the rest of the settings at their default values and then choose **Create**. A new key should appear on the key list page.
```
```suggestion
1. Enter *razorkey* in the **Name** field. Leave the rest of the settings at their default values and then choose **Create**. A new key appears on the key list page.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+## Connect the Azure Services
+
+The Container App requires a secure connection to the storage account and the key vault services in order to resolve the data protection errors and scale correctly. You can connect your services together using the following steps:
```
```suggestion
The Container App requires a secure connection to the storage account and the key vault services in order to resolve the data protection errors and scale correctly. The new services are connected together using the following steps:
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+The Container App requires a secure connection to the storage account and the key vault services in order to resolve the data protection errors and scale correctly. You can connect your services together using the following steps:
+
+> [!IMPORTANT]
+> Security role assignments through Service Connector and other tools generally take a minute or two to propagate, and in some rare cases can take up to eight minutes.
```
```suggestion
> Security role assignments through Service Connector and other tools typically take a minute or two to propagate, and in some rare cases can take up to eight minutes.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+#### Connect the storage account
+
+1. In the Azure portal, navigate to your Container App overview page.
```
```suggestion
1. In the Azure portal, navigate to the Container App overview page.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. In the Azure portal, navigate to your Container App overview page.
+1. On the left navigation, select **Service connector**
+1. On the Service Connector page, choose **+ Create** to open the **Creation Connection** flyout panel and enter the following values:
+    * **Container**: Select the Container App you created previously.
```
```suggestion
    * **Container**: Select the Container App created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. On the Service Connector page, choose **+ Create** to open the **Creation Connection** flyout panel and enter the following values:
+    * **Container**: Select the Container App you created previously.
+    * **Service type**: Choose **Storage - blob**.
+    * **Subscription**: Select the subscription you used previously.
```
```suggestion
    * **Subscription**: Select the subscription previously used .
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Container**: Select the Container App you created previously.
+    * **Service type**: Choose **Storage - blob**.
+    * **Subscription**: Select the subscription you used previously.
+    * **Connection name**: Leave the default value or enter a name of your choosing.
```
same as before

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Service type**: Choose **Storage - blob**.
+    * **Subscription**: Select the subscription you used previously.
+    * **Connection name**: Leave the default value or enter a name of your choosing.
+    * **Storage account**: Select the storage account you created earlier.
```
```suggestion
    * **Storage account**: Select the storage account created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Select **Next: Authentication** to progress to the next step.
+1. Select **System assigned managed identity** and choose **Next: Networking**.
+1. Leave the default networking options selected, and then choose **Review + Create**.
+1. After Azure validates your settings, select **Create**.
```
```suggestion
1. After Azure validates the settings, select **Create**.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. In the Azure portal, navigate to your Container App overview page.
+1. On the left navigation, select **Service connector**.
+1. On the Service Connector page, choose **+ Create** to open the **Creation Connection** flyout panel and enter the following values:
+    * **Container**: Select the container app you created previously.
```
```suggestion
    * **Container**: Select the container app created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Container**: Select the container app you created previously.
+    * **Service type**: Choose **Key Vault**.
+    * **Subscription**: Select the subscription you used previously.
+    * **Connection name**: Leave the default value or enter a name of your choosing.
```
same

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Service type**: Choose **Key Vault**.
+    * **Subscription**: Select the subscription you used previously.
+    * **Connection name**: Leave the default value or enter a name of your choosing.
+    * **Key vault**: Select the key vault you created earlier.
```
```suggestion
    * **Key vault**: Select the key vault created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Select **Next: Authentication** to progress to the next step.
+1. Select **System assigned managed identity** and choose **Next: Networking**.
+1. Leave the default networking options selected, and then choose **Review + Create**.
+1. After Azure validates your settings, select **Create**.
```
```suggestion
1. After Azure validates the settings, select **Create**.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+## Configure and redeploy the app
+
+The necessary Azure resources have been created, so you'll need to configure your app code to point to those services.
```
```suggestion
The necessary Azure resources have been created. In this section the app code is configured to use the new resources.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
````diff
@@ -0,0 +1,320 @@
+    * **Azure.Extensions.AspNetCore.DataProtection.Blobs**: Allows storing ASP.NET Core DataProtection keys in Azure Blob Storage so that keys can be shared across several instances of a web app.
+    * **Azure.Extensions.AspNetCore.DataProtection.Keys**: Enables protecting keys at rest using the Azure Key Vault Key Encryption/Wrapping feature.
+
+    ```dotnetcli
````
Normally we tell them to do this in VS, but this is probably OK. See [this sample](https://learn.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/model?view=aspnetcore-6.0&tabs=visual-studio#scaffold-the-movie-model) if you want to use VS instructions in the future.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+    :::code language="csharp" source="~/../AspNetCore.Docs.Samples/tutorials/scalable-razor-apps/end/Program.cs" id="snippet_ProgramConfigurations" highlight="1-4,6-7,13-19":::
+
+These changes allow the app to manage data protection using a centralized, scalable architecture. `DefaultAzureCredential` discovers the managed identity configurations you enabled earlier when the app is redeployed to Azure.
```
```suggestion
The preceding changes allow the app to manage data protection using a centralized, scalable architecture. `DefaultAzureCredential` discovers the managed identity configurations enabled earlier when the app is redeployed.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+These changes allow the app to manage data protection using a centralized, scalable architecture. `DefaultAzureCredential` discovers the managed identity configurations you enabled earlier when the app is redeployed to Azure.
+
+You'll also need to update the placeholders in `AzureURIs` section of the `appsettings.json` file to include the following:
```
```suggestion
Update the placeholders in `AzureURIs` section of the `appsettings.json` file to include the following:
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+You'll also need to update the placeholders in `AzureURIs` section of the `appsettings.json` file to include the following:
+
+1. Replace the `<storage-account-name>` placeholder with the name of the `scalablerazorstorageXXXX` account you created.
```
```suggestion
1. Replace the `<storage-account-name>` placeholder with the name of the `scalablerazorstorageXXXX` account created.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+You'll also need to update the placeholders in `AzureURIs` section of the `appsettings.json` file to include the following:
+
+1. Replace the `<storage-account-name>` placeholder with the name of the `scalablerazorstorageXXXX` account you created.
+1. Replace the `<key-vault-name>` placeholder with the name of the `scalablerazorvaultXXXX` key vault you created.
```
```suggestion
1. Replace the `<key-vault-name>` placeholder with the name of the `scalablerazorvaultXXXX` key vault created.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+1. Replace the `<storage-account-name>` placeholder with the name of the `scalablerazorstorageXXXX` account you created.
+1. Replace the `<key-vault-name>` placeholder with the name of the `scalablerazorvaultXXXX` key vault you created.
+1. Replace the `<key-name>` placeholder in the key vault URI with the `razorkey` name you created earlier.
```
```suggestion
1. Replace the `<key-name>` placeholder in the key vault URI with the `razorkey` name created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+#### Redeploy the app
+
+Your app is now configured correctly to use the Azure services you created previously. Redeploy the app for your code changes to be applied.
```
```suggestion
The app is now configured correctly to use the Azure services created previously. Redeploy the app for the code changes to be applied.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Right click on the project node in the solution explorer and select **Publish**.
+1. On the publishing profile summary view, select the **Publish** button in the upper right corner.
+
+Visual Studio redeploys the app to the container apps environment you created earlier. When the processes finished, the browser launches to the app home page.
```
```suggestion
Visual Studio redeploys the app to the container apps environment created previously. When the processes finished, the browser launches to the app home page.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+Visual Studio redeploys the app to the container apps environment you created earlier. When the processes finished, the browser launches to the app home page.
+
+Test the app again by searching for *Microsoft* in the search field. The page should now reload with the correct results every time you submit!
```
```suggestion
Test the app again by searching for *Microsoft* in the search field. The page should now reload with the correct results every time you submit.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+## Configure roles for local development
+
+The existing code and configuration of your app can also work while running locally during development. The `DefaultAzureCredential` class you configured earlier is able to pick up local environment credentials to authenticate to Azure Services. You'll need to assign the same roles to your own account that were assigned to your app's managed identity in order for the authentication to work. This  ...
```
```suggestion
The existing code and configuration of the app can also work while running locally during development. The `DefaultAzureCredential` class configured previously is able to pick up local environment credentials to authenticate to Azure Services. You'll need to assign the same roles to your own account that were assigned to your app's managed identity in order for the authentication to work. This should be the same account you use to log into Visual Studio or the Azure CLI.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+#### Assign roles to your developer account
+
+1. In the Azure portal, navigate to the `scalablerazor****` storage account you created earlier.
```
```suggestion
1. In the Azure portal, navigate to the `scalablerazor****` storage account created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Select **Access Control (IAM)** from the left navigation.
+1. Choose **+ Add** and then **Add role assignment** from the drop-down menu.
+1. On the **Add role assignment** page, search for `Storage blob data contributor`, select the matching result, and then choose **Next**.
+1. Make sure **User, group, or service principal** is select, and then choose **+ Select members**.
```
```suggestion
1. Make sure **User, group, or service principal** is selected, and then choose **+ Select members**.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. On the **Add role assignment** page, search for `Storage blob data contributor`, select the matching result, and then choose **Next**.
+1. Make sure **User, group, or service principal** is select, and then choose **+ Select members**.
+1. In the **Select members** flyout, search for your own *user@domain* account and select it from the results.
+1. Choose **Next** and then choose **Review + assign**. After Azure validates your settings, select **Review + assign** again.
```
```suggestion
1. Choose **Next** and then choose **Review + assign**. After Azure validates the settings, select **Review + assign** again.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. In the **Select members** flyout, search for your own *user@domain* account and select it from the results.
+1. Choose **Next** and then choose **Review + assign**. After Azure validates your settings, select **Review + assign** again.
+
+Remember, role assignment permissions make take a minute or two to propagate, or in rare cases up to eight minutes.
```
```suggestion
As previously stated, role assignment permissions might take a minute or two to propagate, or in rare cases up to eight minutes.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+Remember, role assignment permissions make take a minute or two to propagate, or in rare cases up to eight minutes.
+
+You'll also need to repeat the previous steps to assign a role to your account so that it can access the key vault service and secret.
```
```suggestion
Repeat the previous steps to assign a role to your account so that it can access the key vault service and secret.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+
+You'll also need to repeat the previous steps to assign a role to your account so that it can access the key vault service and secret.
+
+1. In the Azure portal, navigate to the `razorscalingkeys` key vault you created earlier.
```
```suggestion
1. In the Azure portal, navigate to the `razorscalingkeys` key vault created previously.
```

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Select **Access Control (IAM)** from the left navigation.
+1. Choose **+ Add** and then **Add role assignment** from the drop-down menu.
+1. On the **Add role assignment** page, search for `Key Vault Crypto Service Encryption User`, select the matching result, and then choose **Next**.
+1. Make sure **User, group, or service principal** is select, and then choose **+ Select members**.
```
```suggestion
1. Make sure **User, group, or service principal** is selected, and then choose **+ Select members**.
```

Rick-Anderson reviewed (changes requested):
Looks like this is VS only so ignore comments about using CLI too. These are all suggestions so if you feel strongly, ignore them. This looks like a great tutorial, I'm anxious to get it published.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
````diff
@@ -0,0 +1,320 @@
+You can use the GitHub Explorer sample app to follow along with this tutorial. Clone the app from GitHub using the following command:
+
+```dotnetcli
+git clone "https://github.com/dotnet/AspNetCore.Docs.Samples.git"
````
This works OK because the [AspNetCore.Docs.Samples](https://github.com/dotnet/AspNetCore.Docs.Samples) is small. @tdykstra  or @guardrex  might know where/how to create a repo for just this sample to make it even simpler.

tdykstra replied:
> @tdykstra or @guardrex might know where/how to create a repo for just this sample to make it even simpler.

Sorry, I don't know where to create a single-sample repo either.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+You'll use Visual Studio to deploy the app to Azure Container Apps. Container apps provide a managed service designed to simplify hosting containerized apps and microservices.
+
+1. Inside the Visual Studio solution explorer, right click on the top level project node and select **Publish**.
+1. In the publishing dialog, select **Azure** as the deployment target, and then choose **Next**.
```
For the style guide, do a global change of `then choose` and replace with `then select`.

alexwolfmsft replied:
Done

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. Inside the Visual Studio solution explorer, right click on the top level project node and select **Publish**.
+1. In the publishing dialog, select **Azure** as the deployment target, and then choose **Next**.
+1. For the specific target, select **Azure Container Apps (Linux)**, and then choose **Next**.
+1. You'll need to create a new container app to deploy to. Select the green **+** icon to open a new dialog and enter the following values:
```
Don't you need to tell them on RG and container apps env they need to select new?

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    * **Resource group**: Select **New** and create a new resource group called *msdocs-scalable-razor*.
+    * **Container apps environment**: Select **New** to open the container apps environment dialog and enter the following values:
+        * **Environment name**: Leave the default value or enter a name of your choosing.
+        * **Location**: Select a location near you.
```
Maybe we should make a statement up front about location. Off the top of my head, something like
Many of the resources created in the app require a location. For this app, location isn't important. A real app should select a location closest to the clients. You may want to select a location near you.

Some of the resources have locations that others don't, so my suggestion to use the previous location won't work.

alexwolfmsft replied:
Yes, I was hesitant to pick a specific location and then always refer to that same one since I know Container Apps and some other services aren't as widely available across regions as others. I'll try to add some kind of note about it.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+        * Select **Ok** to close the container apps environment dialog.
+    * Select **Create** to close the original container apps dialog and Visual Studio creates the container app resource in Azure.
+1. Once the resource is created, make sure it's selected in the list of container apps, and then choose **Next**.
+1. You'll need to create an Azure Container Registry to store the published image artifact for your app. Select the green **+** icon on the container registry screen. Leave the default values, and then select **Create**.
```
Need a red box, 2 pixels wide, around Create in the image.

alexwolfmsft replied:
Done

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+1. You'll need to create an Azure Container Registry to store the published image artifact for your app. Select the green **+** icon on the container registry screen. Leave the default values, and then select **Create**.
+
+    :::image type="content" source="./media/scaling-new-registry-small.png" lightbox="./media/scaling-new-registry.png" alt-text="A screenshot showing how to create a new container registry.":::
+
```
<img width="294" alt="image" src="https://user-images.githubusercontent.com/3605364/190838346-a6d4be0a-eaef-4258-85f1-bd4c10a5f250.png">

Do we need to warn about that?

alexwolfmsft replied:
Done

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
````diff
@@ -0,0 +1,320 @@
+git clone "https://github.com/dotnet/AspNetCore.Docs.Samples.git"
+```
+
+Navigate to the `/tutorials/scalable-razor-apps/start` folder and double select the `ScalableRazor.csproj` file to open the project in Visual Studio.
````
There's no form for submission. Did you forget to update the Index.cshtml file?

alexwolfmsft replied:
Good catch, reviewed the sample again and double checked anything missing.

alexwolfmsft commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+    :::image type="content" source="./media/scaling-deploy-visual-studio-small.png" lightbox="./media/scaling-deploy-visual-studio.png" alt-text="A screenshot showing Visual Studio deployment.":::
+
+    * **Container app name**: Leave the default value or enter a name of your choosing.
+    * **Subscription name**: Select the subscription you'd like to deploy to.
```
Probably, but I'm not sure. Three of them show up for me, I have two personal subscriptions and one MSFT.

alexwolfmsft commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,320 @@
+---
+# Deploying and scaling an ASP.NET Core app on Azure Container Apps
+
+Apps deployed to Azure that experience intermittent high demand benefit from scalability to meet demand. Scalable apps can scale out to ensure capacity during workload peaks and return to normal automatically when the peak drops. Horizontal scaling (scaling out) adds new instances of a resource, such as VMs or database replicas. This article demonstrates how to deploy a horizontally scalable ASP. ...
```
Done

alexwolfmsft commented:
@Rick-Anderson I believe I have implemented all of your feedback. I performed a batch commit of the smaller changes and then some additional work for the other suggestions. Thanks for the thorough review, this has been helpful for learning the voice/standards of this repo.

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,328 @@
+
+Use the search form to browse for GitHub repositories by name.
+
+:::image type="content" source="./media/scaling-app-screenshot.png" alt-text="A screenshot showing the GitHub Explorer app.":::
```
<img width="355" alt="image" src="https://user-images.githubusercontent.com/3605364/191112051-49e75c30-bfb2-4fe2-a634-8d47b531a929.png">

Rick-Anderson commented on aspnetcore/host-and-deploy/scaling-aspnet-apps/scaling-aspnet-apps.md:
```diff
@@ -0,0 +1,328 @@
+
+    :::image type="content" source="./media/scaling-create-new.png" alt-text="A screenshot showing how to create a new container registry.":::
+
+1. Leave the default values, and then select **Create**.
```
<img width="391" alt="image" src="https://user-images.githubusercontent.com/3605364/191112236-c93ef3b4-b8d2-499b-b9d0-a5050714e76a.png">

Rick-Anderson replied:
Looks like a bunch of images are incorrect. BTW, we prefer to use absolute links rather than relative. With absolute links, we can copy/paste anywhere in the repo. We can refactor the MD and move them to different directories.

alexwolfmsft replied:
Neither of these show as broken for me - could this be a browser issue?

![image](https://user-images.githubusercontent.com/93200798/191113269-54e86dd8-e6e7-40e8-9500-f666913202ae.png)

alexwolfmsft replied:
Sometimes images in the preview builds show as broken for me until I refresh a few times.
